src/server/img_storage.rs

```diff
@@ -1,21 +1,77 @@
 //! Fetch and store a given remote image into Google Storage for CDN caching
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+///
+use std::io::Cursor;
+
 use actix_http::http::HeaderValue;
 use actix_web::http::uri;
+use actix_web::web::Bytes;
+use chrono;
 use cloud_storage::{
-    bucket::{Binding, IamPolicy, IamRole, StandardIamRole},
+    bucket::{Binding, IamPolicy, IamRole, RetentionPolicy, StandardIamRole},
     Bucket, Object,
 };
-use serde::Deserialize;
+use image::{io::Reader as ImageReader, ImageFormat};
+use serde::{Deserialize, Serialize};
 
 use crate::error::{HandlerError, HandlerErrorKind, HandlerResult};
 use crate::settings::Settings;
+use crate::tags::Tags;
 
 /// These values generally come from the Google console for Cloud Storage.
-#[derive(Clone, Debug, Deserialize)]
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ImageMetricSettings {
+    /// maximum length of the image
+    max_size: u64,
+    /// max image height
+    max_height: u64,
+    max_width: u64,
+    min_height: u64,
+    min_width: u64,
+    symmetric: bool,
+}
+
+impl Default for ImageMetricSettings {
+    fn default() -> Self {
+        Self {
+            max_size: 100_000,
+            max_height: 256,
+            max_width: 256,
+            min_height: 96,
+            min_width: 96,
+            symmetric: true,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
 pub struct StorageSettings {
+    /// The GCP Cloud storage project name
     project_name: String,
+    /// The Bucket name for this data
     bucket_name: String,
-    endpoint: String,
+    /// The external CDN host
+    #[serde(default = "default_cdn")]
+    cdn_host: String,
+    /// The bucket TTL is determined by the policy set for the given bucket when it's created.
+    #[serde(default = "default_ttl")]
+    bucket_ttl: u64,
+    /// The max time to live for cached data, ~ 15 days.
+    #[serde(default = "default_ttl")]
+    cache_ttl: u64,
+    /// Max dimensions for an image
+    #[serde(default = "ImageMetricSettings::default")]
+    metrics: ImageMetricSettings,
+}
+
+fn default_ttl() -> u64 {
```

Comment:
nit: since StorageSettings has a `Default` impl we shouldn't need these functions and their associated `#[serde(default = "default_ttl")]`.

Instead we can use the `#[serde(default)]` container attribute on the entire struct (see above) to fill these in from the `Default` impl.